test/integration/daemonset/util.go

```diff
@@ -0,0 +1,522 @@
+		pod, err = podClient.Update(newPod)
+		return err
+	}); err != nil {
+		t.Fatalf("failed to update status of pod %q: %v", pod.Name, err)
+	}
+	return pod
+}
+
+func updateHistory(t *testing.T, controllerRevisionClient appstyped.ControllerRevisionInterface, historyName string, updateFunc func(*apps.ControllerRevision)) *apps.ControllerRevision {
+	var history *apps.ControllerRevision
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newHistory, err := controllerRevisionClient.Get(historyName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		updateFunc(newHistory)
+		history, err = controllerRevisionClient.Update(newHistory)
+		return err
+	}); err != nil {
+		t.Fatalf("failed to update status of history %q: %v", history.Name, err)
+	}
+	return history
+}
+
+// deleteDaemonSetAndOrphanPods deletes the given DaemonSet and orphans all its dependents.
+// It also checks that all dependents are orphaned, and the DaemonSet is deleted.
+func deleteDaemonSetAndOrphanPods(
+	dsClient appstyped.DaemonSetInterface,
+	podClient corev1typed.PodInterface,
+	controllerRevisionClient appstyped.ControllerRevisionInterface,
+	podInformer cache.SharedIndexInformer,
+	ds *apps.DaemonSet,
+	t *testing.T) {
+
+	ds, err := dsClient.Get(ds.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("failed to get daemonset %q: %v", ds.Name, err)
+	}
+
+	deletePropagationOrphanPolicy := metav1.DeletePropagationOrphan
+	deleteOptions := &metav1.DeleteOptions{
+		PropagationPolicy: &deletePropagationOrphanPolicy,
+		Preconditions:     metav1.NewUIDPreconditions(string(ds.UID)),
+	}
+	if err = dsClient.Delete(ds.Name, deleteOptions); err != nil {
+		t.Fatalf("failed deleting daemonset %q: %v", ds.Name, err)
+	}
+
+	checkDaemonSetDeleted(dsClient, ds.Name, t)
+	checkDaemonSetPodsOrphaned(podClient, t)
+	checkDaemonSetHistoryOrphaned(controllerRevisionClient, t)
+}
+
+func checkDaemonSetPodsName(podInformer cache.SharedIndexInformer, podNamePrefix string, t *testing.T) {
+	pods := podInformer.GetIndexer().List()
+	for _, object := range pods {
+		pod := object.(*v1.Pod)
+		if !strings.HasPrefix(pod.Name, podNamePrefix) {
+			t.Fatalf("expected pod %q has name prefix %q", pod.Name, podNamePrefix)
+		}
+	}
+}
+
+func waitDaemonSetAdoption(
+	podClient corev1typed.PodInterface,
+	controllerRevisionClient appstyped.ControllerRevisionInterface,
+	podInformer cache.SharedIndexInformer,
+	ds *apps.DaemonSet,
+	podNamePrefix string,
+	t *testing.T) {
+	checkDaemonSetPodsAdopted(podClient, podInformer, ds.UID, t)
+	checkDaemonSetHistoryAdopted(controllerRevisionClient, ds.UID, t)
+
+	// Ensure no pod is re-created by checking their names
+	checkDaemonSetPodsName(podInformer, podNamePrefix, t)
+}
+
+// setupGC starts a GC for given test server. The GC is only started when workerCount > 0.
+// The function returns a tear down function to defer shutting down the GC.
+func setupGC(t *testing.T, server *httptest.Server, workerCount int) func() {
```

Comment:
We can simplify this for the integration test - we don't need a workerCount parameter I think - just have a fixed value.

Reply:
Agreed, just set it to a fixed value suffices. We can parameterize some of the GC settings if needed in future, but it is unnecessary now.